### PR TITLE
[mesheryctl] `system start` command bug of deployment of config components adapters

### LINKED ISSUE #9161: [mesheryctl] adapters aren't deployed specified in context's components field (docker)
_See [mesheryctl Command Tracker](https://bit.ly/3dqXy1q) for current status of commands._

### Current Behavior
when  meshery is started using `mesheryctl system start` command, adapters specified in the `components` field of the current context in `.meshery/config.yaml` file, aren't being deployed.

### What's Bug

when the user executes `mesheryctl system start`, [docker-compose.yaml](https://github.com/meshery/meshery/blob/b9c5c8b76d9a55a4a74ac385164a1a6511f03c92/install/docker/docker-compose.yaml) file is downloaded and deployed. [docker-compose.yaml](https://github.com/meshery/meshery/blob/b9c5c8b76d9a55a4a74ac385164a1a6511f03c92/install/docker/docker-compose.yaml) file has only two main services specified in it i.e. `meshery` and `watchtower`, so adapters of components field aren't dynamically added to the docker-compose file.

### Environment

- Host OS: Mac Linux Windows
- Meshery Server Version: stable-v
- Meshery Client Version: stable-v
- Platform: Docker Kubernetes

---

### Contributor [Guides](https://docs.meshery.io/project/contributing) and [Handbook](https://layer5.io/community/handbook)
- 📚 [mesheryctl Contributing Guide](https://github.com/meshery/meshery/blob/master/mesheryctl/README.md)
- 📺 [Beginner's guide to contributing to Meshery and mesheryctl](https://youtu.be/hh_kFLZx3G4)
- ⌨️ [mesheryctl Command Tracker](https://docs.google.com/spreadsheets/d/1q63sIGAuCnIeDs8PeM-0BAkNj8BBgPUXhLbe1Y-318o/edit#gid=0)
- ⌨️ [Meshery CLI Commands and Documentation](https://docs.google.com/document/d/1xRlFpElRmybJ3WacgPKXgCSiQ2poJl3iCCV1dAalf0k/edit#heading=h.5fucij4hc5wt)
- 🙋🏾🙋🏼 Questions: [Discussion Forum](http://discuss.meshery.io) and [Community Slack](https://slack.meshery.io)

### PR DESCRIPTION
**Pull request by zakisk**
**Notes for Reviewers**

This PR fixes #9161

**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

File: mesheryctl/internal/cli/root/system/start.go [hunk 1]
```diff
@@ -221,13 +221,20 @@ func start() error {
 
 			temp, ok := utils.Services[v]
 			if !ok {
-				return errors.New("unable to extract component version")
+				return errors.New(fmt.Sprintf("no such service `%s` exists.", v))
```
Comment (leecalcote) [inline mesheryctl/internal/cli/root/system/start.go:224]: @zakisk if this is user facing, it'd be helpful to expand on this message.
```suggestion
				return errors.New(fmt.Sprintf("No Docker Compose service exists for Meshery component `%s`.", v))
```

File: mesheryctl/internal/cli/root/system/start.go [hunk 2]
```diff
@@ -221,13 +221,20 @@ func start() error {
...
 			spliter := strings.Split(temp.Image, ":")
 			temp.Image = fmt.Sprintf("%s:%s-%s", spliter[0], currCtx.GetChannel(), "latest")
 			utils.Services[v] = temp
 			AllowedServices[v] = utils.Services[v]
+			utils.ViperCompose.Set(fmt.Sprintf("services.%s", v), utils.Services[v])
+			err = utils.ViperCompose.WriteConfig()
+			if err != nil {
+				// failure while adding a service to docker compose file is not a fatal error
+				// mesheryctl will continue deploying with required services (meshery, watchtower)
+				log.Infof("Encountered an error while adding `%s` service to Docker compose file, you may not have permission to write file.", v)
```
Comment (leecalcote) [inline mesheryctl/internal/cli/root/system/start.go:236]:
```suggestion
				log.Infof("Encountered an error while adding `%s` service to Docker Compose file. Verify permission to write to file.", v)
```

Which "file" - - this should be info in the log, too.

This should be using a MeshKit-style error with well-formed error code and cause and remediation.

**Review by leecalcote — CHANGES_REQUESTED** [review]

**Comment by leecalcote** [comment]
Thanks for this, @zakisk. The lint errors are piling up...

**Comment by leecalcote** [comment]
It'd be very helpful, if you could take a look at the MeshKit Error Codes as well.

**Comment by saurabh100ni** [comment]
Let's discuss this on the Meshery dev call tomorrow at 7:30 pm IST / 9:00 am CT,
Please add this as your agenda item in Docs: https://docs.google.com/document/d/1Dg4Uz9rGTuGHmYZwIt47z1j5x3LmnpxD1M1SgypAkQ0/edit

**Comment by leecalcote** [comment]
Syncing branch...

File: mesheryctl/internal/cli/root/system/start.go [hunk 3]
```diff
@@ -221,13 +221,20 @@ func start() error {
...
 			spliter := strings.Split(temp.Image, ":")
 			temp.Image = fmt.Sprintf("%s:%s-%s", spliter[0], currCtx.GetChannel(), "latest")
 			utils.Services[v] = temp
 			AllowedServices[v] = utils.Services[v]
+			utils.ViperCompose.Set(fmt.Sprintf("services.%s", v), utils.Services[v])
+			err = utils.ViperCompose.WriteConfig()
+			if err != nil {
+				// failure while adding a service to docker compose file is not a fatal error
+				// mesheryctl will continue deploying with required services (meshery, watchtower)
+				log.Infof("Encountered an error while adding `%s` service to Docker Compose file. Verify permission to write to `.meshery/meshery.yaml` file.", v)
```
Comment (leecalcote) [inline mesheryctl/internal/cli/root/system/start.go:236]: @MUzairS15 should this be using a MeshKit-style error?

Comment (leecalcote) [reply mesheryctl/internal/cli/root/system/start.go:236]: @zakisk, will you convert this to a MeshKit Error?

Comment (zakisk) [reply mesheryctl/internal/cli/root/system/start.go:236]: @leecalcote at the I kept it as a warning to user that the adapter defined in context won't be deployed due to the error,  so do I break the course of starting meshery by returning Meshkit Error or keep this as warning but format it in Meshkit style?

Comment (leecalcote) [reply mesheryctl/internal/cli/root/system/start.go:236]: Fail forward, please. Keep deploying Meshery, despite the failed adapter.

I think that you made the best decision, @zakisk

**Pull request merged**